### ISSUE TITLE
Question about features and results of the 3 former reference frames while training.

### ISSUE THREAD

**Issue by RaymondWang987** [GuoleiSun/VSS-MRCFA#1]
Thanks for releasing the code. I have one question about the training protocol. In your methods of CFFM(CVPR2022) and MRCFA(ECCV 2022)，the 3 former reference frames are used to generate the results of the last key frame. But I find that yours models output feature maps and generate their segmentation logits  of all the 4 frames by output head (small decoder). Do you use the 4 frames results for supervision while training? If so, how do you generate the features of the former reference frames? I think the process for generating features of reference frames might not be equivelent to the process of the last key frame. How do you deal with them?
Hoping for your reply.
```
out1=resize(self.deco1(outs[0]), size=(h, w),mode='bilinear',align_corners=False).unsqueeze(1)
out2=resize(self.deco2(outs[1]), size=(h, w),mode='bilinear',align_corners=False).unsqueeze(1)
out3=resize(self.deco3(outs[2]), size=(h, w),mode='bilinear',align_corners=False).unsqueeze(1)
# out4=resize(self.deco4(outs[3]+outs[2]+outs[1]+outs[0]), size=(h, w),mode='bilinear',align_corners=False).unsqueeze(1)
# out4=resize(self.deco4((outs[3]+outs[2]+outs[1])/3.0+outs[0]), size=(h, w),mode='bilinear',align_corners=False).unsqueeze(1)
out4=resize(self.deco4((outs[0]+outs[1]+outs[2])/3.0+outs[3]), size=(h, w),mode='bilinear',align_corners=False).unsqueeze(1)
output=torch.cat([x,out1,out2,out3,out4],dim=1)   ## b*(k+k)*124*h*w
if not self.training:
            # return output.squeeze(1)
            # return torch.cat([x2,x3],1).mean(1)
            return out4.squeeze(1)
            # return out4.squeeze(1)+(out3.squeeze(1)+out2.squeeze(1)+out1.squeeze(1))/3
            # return F.softmax(torch.cat([out1,out2,out3,out4],1),dim=2).sum(1)
            # return torch.cat([out1,out2,out3,out4],1).mean(1)
return output
```

**Comment by GuoleiSun** [comment]
Hello, thanks for your interest.
As you may see from our code, the process for generating the features of the reference frames is the same as the one for last target frame, i.e., using the same backbone. The supervision for all 4 frames is used while training. The supervision for the main prediction is the **main loss**, while other losses are **auxiliary losses**.
We can slightly analyse this line of code: output=torch.cat([x,out1,out2,out3,out4],dim=1) 
1. x is the predicted logits for 4 frames, there is no interaction between 4 frames and x is just like the output from the image segmentation network.
2. The features of reference frame 1, 2 and 3 are used to refine the target frame features, denoted as outs[0], outs[1], and outs[3]. So out1, out2, and out3 are actually the predicted logits for the target frame. 
3. out4 is our main prediction for the target frame, which combines all the available information.

Please let me know if you have more questions.

**Comment by RaymondWang987** [comment]
Thanks for your reply! If I understand correct the outs[0], outs[1], and outs[2] are the O<sub>ti</sub>=B<sub>ti</sub><sup>1</sup>*F<sub>ti</sub> in equation (7). Do you mean that the outs[0], outs[1], and outs[2] are not the features of reference frame 1, 2 and 3, but the features of the key frame refined by reference frame 1, 2 and 3? If this case, out1, out2, and out3 are supervised by the target frame but not the reference frame 1, 2 and 3, is that true?

**Comment by RaymondWang987** [comment]
But I think O<sub>ti</sub>=B<sub>ti</sub> <sup>1</sup> * F<sub>ti</sub>, it's accurally the feature of reference frame ti multiplied by a weight B<sub>ti</sub>（calculated by attention on multi scales features). So it's still the feature of the reference frame and supervised by the gt of reference frame ti. I don't know if I understand it right.

**Comment by GuoleiSun** [comment]
> Thanks for your reply! If I understand correct the outs[0], outs[1], and outs[2] are the Oti=Bti1*Fti in equation (7). Do you mean that the outs[0], outs[1], and outs[2] are not the features of reference frame 1, 2 and 3, but the features of the key frame refined by reference frame 1, 2 and 3? If this case, out1, out2, and out3 are supervised by the target frame but not the reference frame 1, 2 and 3, is that true?

This is correct

**Comment by GuoleiSun** [comment]
> But I think Oti=Bti 1 * Fti, it's accurally the feature of reference frame ti multiplied by a weight Bti（calculated by attention on multi scales features). So it's still the feature of the reference frame and supervised by the gt of reference frame ti. I don't know if I understand it right.

Remember that {B_ti}^1 is not a simple weight, you can think attention of (query,key,value), query* key is the attention and query* key can be used to retrieve values from value.

**Comment by RaymondWang987** [comment]
Is CFFM also in this case? I mean the auxiliary losses on out1,out2,out3 supervised by the target frame. Have you ever tried to train the model without these auxiliary losses? How much the auxiliary losses on out1,out2,out3 influence performance?

**Comment by RaymondWang987** [comment]
By the way, how do you set the weight and ratio between the auxiliary losses on out1,out2,out3 and the main loss on out4? Thanks for your detailed reply!

**Comment by GuoleiSun** [comment]
> Is CFFM also in this case? I mean the auxiliary losses on out1,out2,out3 supervised by the target frame. Have you ever tried to train the model without these auxiliary losses? How much the auxiliary losses on out1,out2,out3 influence performance?

CFFM has very different structure and is not in this case. I don't think the auxiliary losses on out1, out2 and out3 influence performance much, but it is intuitive to have them since they are the new representation of the target frame.

**Comment by GuoleiSun** [comment]
> By the way, how do you set the weight and ratio between the auxiliary losses on out1,out2,out3 and the main loss on out4? Thanks for your detailed reply!

We don't specifically tune the weight, and use 1.0 for out1, out2, out3 and out4.

**Comment by RaymondWang987** [comment]
Thanks for your reply! It helps me a lot.
